borsh/src/schema_helpers.rs

```diff
@@ -36,3 +36,270 @@ pub fn schema_container_of<T: BorshSchema>() -> BorshSchemaContainer {
+        }
+
+        // Tuples and structs sum sizes of all the members.
+        Ok(Definition::Tuple { elements }) => tuple(count, elements, defs, stack),
+        Ok(Definition::Struct { fields }) => match fields {
+            Fields::NamedFields(fields) => tuple(
+                count,
+                fields.into_iter().map(|(_, field)| field),
+                defs,
+                stack,
+            ),
+            Fields::UnnamedFields(fields) => tuple(count, fields, defs, stack),
+            Fields::Empty => Ok(0),
+        },
+
+        // Primitive types.
+        Err("nil") => Ok(0),
+        Err("bool" | "i8" | "u8") => Ok(count),
+        Err("i16" | "u16") => mul(count, 2),
+        Err("i32" | "u32" | "f32") => mul(count, 4),
+        Err("i64" | "u64" | "f64") => mul(count, 8),
+        Err("i128" | "u128") => mul(count, 16),
+
+        // string is just Vec<u8>
+        Err("string") => mul(count, add(MAX_LEN, 4)?),
+
+        Err(_) => Err(MaxSizeError::MissingDefinition),
+    }?;
+
+    stack.pop();
+    Ok(res)
+}
+
+#[test]
+fn test_max_serialized_size() {
+    use alloc::string::{String, ToString};
+
+    const MAX_LEN: usize = u32::MAX as usize;
+
+    assert_eq!(Ok(2), max_serialized_size::<u16>());
+    assert_eq!(Ok(8), max_serialized_size::<usize>());
+
+    assert_eq!(Ok(1), max_serialized_size::<Option<()>>());
+    assert_eq!(Ok(2), max_serialized_size::<Option<u8>>());
+    assert_eq!(
+        Ok(9),
+        max_serialized_size::<core::result::Result<u8, usize>>()
+    );
+
+    assert_eq!(Ok(0), max_serialized_size::<()>());
+    assert_eq!(Ok(1), max_serialized_size::<(u8,)>());
+    assert_eq!(Ok(5), max_serialized_size::<(u8, u32)>());
+
+    assert_eq!(Ok(0), max_serialized_size::<[u8; 0]>());
+    assert_eq!(Ok(16), max_serialized_size::<[u8; 16]>());
+    assert_eq!(Ok(16), max_serialized_size::<[[u8; 4]; 4]>());
+
+    assert_eq!(Ok(4 + MAX_LEN), max_serialized_size::<Vec<u8>>());
+    assert_eq!(Ok(4 + MAX_LEN), max_serialized_size::<String>());
+
+    assert_eq!(
+        Err(MaxSizeError::Overflow),
+        max_serialized_size::<Vec<Vec<u8>>>()
+    );
+    assert_eq!(Ok(4 + MAX_LEN * 4), max_serialized_size::<Vec<Vec<()>>>());
+    assert_eq!(
+        Ok(0),
+        max_serialized_size::<[[[(); MAX_LEN]; MAX_LEN]; MAX_LEN]>()
+    );
+
+    use crate as borsh;
+
+    #[derive(::borsh_derive::BorshSchema)]
+    pub struct Empty;
+
+    #[derive(::borsh_derive::BorshSchema)]
+    pub struct Named {
+        _foo: usize,
+        _bar: [u8; 15],
+    }
+
+    #[derive(::borsh_derive::BorshSchema)]
+    pub struct Unnamed(usize, [u8; 15]);
+
+    #[derive(::borsh_derive::BorshSchema)]
+    struct Multiple {
+        _usz0: usize,
+        _usz1: usize,
+        _usz2: usize,
+        _vec0: Vec<usize>,
+        _vec1: Vec<usize>,
+    }
+
+    #[derive(::borsh_derive::BorshSchema)]
+    struct Recursive(Option<alloc::boxed::Box<Recursive>>);
+
```

Comment:
add a test for a prototype `BoundedVec<T, L, U>`. 
it should be empty, just have a `BorshSchema` implementation